anvill/include/anvill/Lifters/Context.h

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2021 Trail of Bits, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <optional>
+
+namespace llvm {
+class Constant;
+class GlobalValue;
+class Module;
+}  // namespace llvm
+namespace remill {
+class Arch;
+}  // namespace remill
+namespace anvill {
+
+class ContextImpl;
+class FunctionLifter;
+class LifterOptions;
+class MemoryProvider;
+class TypeProvider;
+class ValueLifter;
+
+// Lifting context for ANVILL. The lifting context keeps track of the options
+// used for lifting, the module into which lifted objects are placed, and
+// a the mapping between lifted objects and their original addresses in the
+// binary.
+class Context {
+ public:
+  ~Context(void);
+
+  explicit Context(const LifterOptions &options,
+                   const std::shared_ptr<MemoryProvider> &mem_provider_,
+                   const std::shared_ptr<TypeProvider> &type_provider_);
+
+  // Assuming that `entity` is an entity that was lifted by this `EntityLifter`,
+  // then return the address of that entity in the binary being lifted.
+  std::optional<uint64_t> AddressOfEntity(llvm::GlobalValue *entity) const;
+
+  // Return the options being used by this entity lifter.
+  const LifterOptions &Options(void) const;
+
+  Context(const Context &) = default;
```

Comment:
Same as with the destructor. Do we need to declare all the default constructors and assignment operators explicitly? Aside from the private one.

Reply:
Mostly I see it as a way of declaring what I intend to be valid behaviours. Omitting it means that you get default behaviour if it is possible, but it doesn't mean your API intends for the default behaviour to actually be used. Making it explicit communicates that you thought of how your classes can be constructed/copied, and declared to the world that the default implementation is in fact good. 

Really, we should use a macro to pump out these repetitive types of defaults, or repetitive deletes.

Reply:
Honestly, if I had to pick between a macro and a repetitive piece of code, I'd take the code, since it's easier to read and I don't have to look for the macro definition or think about how the macro is going to unfold. As for communicating intent, the absence of assignment operators and copy constructors means that default behavior is "explicitly" intended, IMO. Hasn't rule-of-zero become sort of a standard?